### PR TITLE
Implement LORM as described

### PR DESCRIPTION
**Pull request by ilyakharlamov**
Currently, LORM metric is not implemented and not tested
although it seems to do something but definitely doesn't do what's
described in the paper and claimed in the description : (LORM = (R /
RN) where RN = N * (N - 1) / 2, N = total number of methods, R - method
relationships)

This PR:

- adds two unit tests for LORM: one as a parameter to the common
MetricsTest, another one for the specific metric LormTest
- implements the LORM metric according to the paper and the description

### CODE REVIEW

**Comment by 0crat** [comment]
@paulodamaso[/z](https://www.0crat.com/u/paulodamaso) please, pay attention to this pull request

**Comment by 0crat** [comment]
Job [#292](https://github.com/yegor256/jpeek/issues/292) is now in scope, role is `REV`

**Comment by 0crat** [comment]
@paulodamaso[/z](https://www.0crat.com/u/paulodamaso) everybody who has role `REV` is banned at [#292](https://github.com/yegor256/jpeek/issues/292); I won't be able to assign anyone automatically; consider assigning someone manually (as in [§19](http://www.zerocracy.com/policy.html#19)), or invite more people (as in [§51](http://www.zerocracy.com/policy.html#51)), or remove the job from the scope (as in [§14](http://www.zerocracy.com/policy.html#14))

File: src/test/java/org/jpeek/metrics/LormTest.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                new FakeBase("TwoCommonMethods")
+            ).xml()
+        );
+        final int methods = 6;
+        MatcherAssert.assertThat(
+            "N variable",
+            xml.xpath(
+                "//class[@id='TwoCommonMethods']/vars/var[@id='N']/text()"
+            ).get(0),
+            Matchers.equalTo(
```
Comment (paulodamaso) [inline src/test/java/org/jpeek/metrics/LormTest.java]: @ilyakharlamov Let's avoid using static matchers. Use `IsEqual`  instead.

File: src/test/java/org/jpeek/metrics/LormTest.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                String.valueOf(methods)
+            )
+        );
+        MatcherAssert.assertThat(
+            "R variable",
+            xml.xpath(
+                "//class[@id='TwoCommonMethods']/vars/var[@id='R']/text()"
+            ).get(0),
+            Matchers.equalTo(
```
Comment (paulodamaso) [inline src/test/java/org/jpeek/metrics/LormTest.java]: @ilyakharlamov Same as above

File: src/test/java/org/jpeek/metrics/LormTest.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                "org/jpeek/metrics/LORM.xsl"
+            ).stream()
+        ).transform(
+            new Skeleton(
+                new FakeBase("TwoCommonMethods")
+            ).xml()
+        );
+        final int methods = 6;
+        MatcherAssert.assertThat(
+            "N variable",
```
Comment (paulodamaso) [inline src/test/java/org/jpeek/metrics/LormTest.java]: @ilyakharlamov This string is supposed to be an error message if the assertion fails, please improve it

File: src/test/java/org/jpeek/metrics/LormTest.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            xml.xpath(
+                "//class[@id='TwoCommonMethods']/vars/var[@id='N']/text()"
+            ).get(0),
+            Matchers.equalTo(
+                String.valueOf(methods)
+            )
+        );
+        MatcherAssert.assertThat(
+            "R variable",
```
Comment (paulodamaso) [inline src/test/java/org/jpeek/metrics/LormTest.java]: @ilyakharlamov Same as above

File: src/test/java/org/jpeek/metrics/LormTest.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                    }.length
+                )
+            )
+        );
+        MatcherAssert.assertThat(
+            "RN variable = N * (N - 1) / 2",
+            xml.xpath(
+                "//class[@id='TwoCommonMethods']/vars/var[@id='RN']/text()"
+            ).get(0),
+            Matchers.equalTo(
```
Comment (paulodamaso) [inline src/test/java/org/jpeek/metrics/LormTest.java]: @ilyakharlamov Use `IsEqual` here

File: src/test/java/org/jpeek/metrics/LormTest.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                        "methodTwo   -> methodOne",
+                        "methodThree -> methodOne",
+                        "methodFive  -> methodFour",
+                        "methodSix   -> methodFour",
+                    }.length
+                )
+            )
+        );
+        MatcherAssert.assertThat(
+            "RN variable = N * (N - 1) / 2",
```
Comment (paulodamaso) [inline src/test/java/org/jpeek/metrics/LormTest.java]: @ilyakharlamov Same as above about the message

**Comment by 0crat** [comment]
@paulodamaso[/z](https://www.0crat.com/u/paulodamaso) everybody who has role `REV` is banned at [#292](https://github.com/yegor256/jpeek/issues/292); I won't be able to assign anyone automatically; consider assigning someone manually (as in [§19](http://www.zerocracy.com/policy.html#19)), or invite more people (as in [§51](http://www.zerocracy.com/policy.html#51)), or remove the job from the scope (as in [§14](http://www.zerocracy.com/policy.html#14))

File: src/test/java/org/jpeek/metrics/LormTest.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+                String.valueOf(methods)
+            )
+        );
+        MatcherAssert.assertThat(
+            "variable 'R' is not calculated correctly",
+            xml.xpath(
+                "//class[@id='TwoCommonMethods']/vars/var[@id='R']/text()"
+            ).get(0),
+            new IsEqual<>(
+                String.valueOf(
```
Comment (paulodamaso) [inline src/test/java/org/jpeek/metrics/LormTest.java]: @ilyakharlamov I think that is more elegant to use `ListOf` here to build this string list

Comment (ilyakharlamov) [reply src/test/java/org/jpeek/metrics/LormTest.java]: Unfortunately, that will not work due to a [bug in jcabi-xml](https://github.com/jcabi/jcabi-xml/issues/149) (List\<String> that is returned by xpath(), is not equal to a List\<String>)

File: src/test/java/org/jpeek/metrics/LormTest.java [hunk 8]
```diff
@@ -83,11 +83,11 @@ public void calculatesVariables() throws IOException {
             )
         );
         MatcherAssert.assertThat(
-            "RN variable = N * (N - 1) / 2",
+            String.format("variable 'RN' is not calculated correctly"),
```
[the remark this answers is not in the document]
Comment (ilyakharlamov) [reply src/test/java/org/jpeek/metrics/LormTest.java]: fixed

File: src/test/java/org/jpeek/metrics/LormTest.java (hunk 7 above)
Comment (paulodamaso) [reply src/test/java/org/jpeek/metrics/LormTest.java]: @ilyakharlamov Well, while this bug is not fixed you could use `new IsEqual(String.valueOf(new ListOf(...).length())))`

**Comment by 0crat** [comment]
@paulodamaso[/z](https://www.0crat.com/u/paulodamaso) everybody who has role `REV` is banned at [#292](https://github.com/yegor256/jpeek/issues/292); I won't be able to assign anyone automatically; consider assigning someone manually (as in [§19](http://www.zerocracy.com/policy.html#19)), or invite more people (as in [§51](http://www.zerocracy.com/policy.html#51)), or remove the job from the scope (as in [§14](http://www.zerocracy.com/policy.html#14))

**Comment by 0crat** [comment]
This pull request [#292](https://github.com/yegor256/jpeek/issues/292) is assigned to @ammaratef45[/z](https://www.0crat.com/u/ammaratef45), here is [why](https://www.0crat.com/footprint/C7JGJ00DP/24f76aa9-5011-410a-97ba-f8dbeae0fe7e); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

**Comment by paulodamaso** [comment]
@0crat assign me

**Comment by 0crat** [comment]
> @0crat assign me ([here](https://github.com/yegor256/jpeek/issues/292#issuecomment-454624185))

@paulodamaso This pull request [#292](https://github.com/yegor256/jpeek/issues/292) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/C7JGJ00DP/e2deb22f-5c97-49bd-999a-3feb49eb2eee); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

**Comment by paulodamaso** [comment]
@ilyakharlamov ping

**Comment by paulodamaso** [comment]
@ilyakharlamov Take a look at [this](https://github.com/yegor256/jpeek/pull/292#discussion_r247738755), please

File: src/test/java/org/jpeek/metrics/LormTest.java (hunk 7 above)
Comment (ilyakharlamov) [reply src/test/java/org/jpeek/metrics/LormTest.java]: @paulodamaso Why you suggest comparing the length while we need to compare the content of the array, not just the length.
in 267bb4f I did `new IsEqual<>(
                new int[] {methods * (methods - 1) / 2}
            )` instead. If you have any better ideas comparing the List, please let me know.

Comment (paulodamaso) [reply src/test/java/org/jpeek/metrics/LormTest.java]: @ilyakharlamov I'm suggesting getting rid of 
´´´
                    new String[] {
                        "methodTwo   -> methodOne",
                        "methodThree -> methodOne",
                        "methodFive  -> methodFour",
                        "methodSix   -> methodFour",
                    }.length
´´´

Comment (ilyakharlamov) [reply src/test/java/org/jpeek/metrics/LormTest.java]: @paulodamaso done in in 16777fa

**Comment by paulodamaso** [comment]
@ilyakharlamov Sorry for not being clear enough. I am asking you to replace
```
                    new String[] {
                        "methodTwo   -> methodOne",
                        "methodThree -> methodOne",
                        "methodFive  -> methodFour",
                        "methodSix   -> methodFour",
                    }.length
```
with 
```
new ListOf(
...
).length()
```

File: src/test/java/org/jpeek/metrics/LormTest.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        );
+        MatcherAssert.assertThat(
+            "variable 'R' is not calculated correctly",
+            new ListOf<>(
+                xml.xpath(
+                    "//class[@id='TwoCommonMethods']/vars/var[@id='R']/text()"
+                ).stream().mapToInt(Integer::valueOf).iterator()
+            ),
+            new IsEqual<>(
+                new ListOf<>(
```
Comment (paulodamaso) [inline src/test/java/org/jpeek/metrics/LormTest.java]: @ilyakharlamov Sorry again, I meant:
```
new ListOf<>(
  "methodTwo   -> methodOne",
  "methodThree -> methodOne",
  "methodFive  -> methodFour",
  "methodSix   -> methodFour"
).length()
```

Comment (ilyakharlamov) [reply src/test/java/org/jpeek/metrics/LormTest.java]: @paulodamaso Now I got it, new ListOf<>(...).size() in 66299fd

**Comment by paulodamaso** [comment]
@ilyakharlamov Yes. I think that just using `xpath('...').get(0)` is better: if for some reason it return an empty list the test will throw an exception and will fail anyways, WDYT?

**Comment by paulodamaso** [comment]
@rultor merge

**Comment by 0crat** [comment]
The job [#292](https://github.com/yegor256/jpeek/issues/292) is now out of scope

**Comment by 0crat** [comment]
Order was finished: +15 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

**Pull request merged**